### PR TITLE
Add `normalize` and `normalize!`

### LINKED ISSUE #246: Add Base.normalize and normalize! to Compat
It would be great to have Base.normalize and normalize! in Compat. Normalization is a very basic and common concept - having it in Compat would allow 0.4-compatible packages to use it and to provide new methods for it. It would be very natural to have `normalize!(h::StatsBase.Histogram)`, for example.

### PR DESCRIPTION
**Pull request by ranjanan**
Add `normalize` and `normalize!`, which normalizes a vector according to the p-norm. Fixes #246 .

### CODE REVIEW

**Comment by ranjanan** [comment]
Failing on v0.3 :( I added docstrings to the functions, but there's no `@doc` defined for v0.3.

**Comment by TotalVerb** [comment]
@ranjanan I think the standard way to deal with docstrings is to use the syntax without `@doc`, that is:

```julia
"""
Docstring
"""
function x() end
```

This degrades to a no-op on v0.3, and will produce the correct docstring on v0.4.

**Comment by ranjanan** [comment]
Yes, I tried that first, but the docs won't show up at `?help`. @yuyichao told me this was by design. That was why I shifted to using `@doc`.

**Comment by TotalVerb** [comment]
It is really no surprise that docs won't show up on v0.3, as docstrings were not implemented until v0.4. The `@doc` macro won't really help, as that was also implemented in v0.4.

The docs should show up on v0.4 without issue using the docstring syntax.

**Comment by ranjanan** [comment]
@TotalVerb I meant that on v0.4, the normal docstring syntax within an `if` block does _not_ work by design. On v0.3 `@doc` wasn't part of Base Julia so it's not going to work. That is why, at @tkelman 's suggestion, I removed it and linked to it in the README.

**Comment by TotalVerb** [comment]
@ranjanan Ah, you are correct. I apologize for the misunderstanding.

File: src/Compat.jl [hunk 1]
```diff
@@ -1085,6 +1085,43 @@ if !isdefined(Base, :Threads)
     export Threads
 end
 
+if !isdefined(Base, :normalize)
+
```
Comment (tkelman) [inline src/Compat.jl]: don't put empty lines at the beginning or end of blocks like this

**Comment by tkelman** [comment]
needs rebasing

**Comment by TotalVerb** [comment]
You might need to provide copy_oftype too.

File: test/runtests.jl [hunk 2]
```diff
@@ -1286,3 +1286,12 @@ let
     @test norm(w) === 1.0
     @test norm(normalize!(v) - w, Inf) < eps()
 end
+
+# 0.5 style single argument `@boundscheck`
```
Comment (tkelman) [inline test/runtests.jl]: rebasing, not merging

**Comment by ranjanan** [comment]
@tkelman , yes, but I screwed up. I rebased on my local branch and my push got rejected, so I  ended merging stuff from my remote branch on my local branch. I need to get better at this.

**Comment by tkelman** [comment]
when you rebase you need to force push over your branch

**Comment by ranjanan** [comment]
So the `norm` test is failing. Which is kinda weird, I [copied this test](https://github.com/JuliaLang/julia/blob/master/test/linalg/generic.jl#L254) from the Julia tests

File: test/runtests.jl [hunk 3]
```diff
@@ -1280,3 +1280,31 @@ end
...
+        end
+    end
+end
+
+#Test potential overflow in normalize!
+let
+    δ = inv(prevfloat(typemax(Float64)))
+    v = [δ, -δ]
+
+    @test norm(v) === 7.866824069956793e-309
```
Comment (ranjanan) [inline test/runtests.jl]: This is where the v0.3 test fails. `norm(v)` gives `Inf`

Comment (ranjanan) [inline test/runtests.jl]: You know what, I'm not sure why I'm adding this test for norm. I'm testing normalize and normalize!

Comment (ranjanan) [inline test/runtests.jl]: Well, actually it seems that `norm` itself behaves differently on v0.3 as compared to v0.4 and v0.5.

**Comment by ranjanan** [comment]
Okay, the function `norm` behaves differently in v0.3 as compared to v0.4 and v0.5. It seems to me that to support `normalize` and `normalize!`, we drop v0.3 support, or warn against flaky behaviour.

**Comment by stevengj** [comment]
At some point the `norm` function's behavior was improved to reduce overflow/underflow problems.   I don't think it's a big deal if `Compat.normalize` has the same limitations as `norm` in a given Julia version.

File: test/runtests.jl [hunk 4]
```diff
@@ -1301,9 +1301,13 @@ end
 let
     δ = inv(prevfloat(typemax(Float64)))
     v = [δ, -δ]
-
+    if VERSION > v"0.4.0-"
```
Comment (stevengj) [inline test/runtests.jl]: `VERSION` >= v"0.4.0-pre+7164", for JuliaLang/julia#1861

File: test/runtests.jl [hunk 5]
```diff
@@ -1301,9 +1301,13 @@ end
...
+        @test norm(v) === 7.866824069956793e-309
+    end
     w = normalize(v)
     @test w ≈ [1/√2, -1/√2]
-    @test norm(w) === 1.0
+    if VERSION > v"0.4.0-"
+        @test norm(w) === 1.0
```
Comment (stevengj) [inline test/runtests.jl]: This seems too strict.  Should be `@test norm(w) ≈ 1.0`, and this should work on any version of Julia.

File: test/runtests.jl [hunk 6]
```diff
@@ -1280,3 +1280,32 @@ end
...
+let
+    δ = inv(prevfloat(typemax(Float64)))
+    v = [δ, -δ]
+    if VERSION > v"0.4.0-"
+        @test norm(v) === 7.866824069956793e-309
+    end
+    w = normalize(v)
+    @test w ≈ [1/√2, -1/√2]
+    @test_approx_eq norm(w) 1.0
+    @test norm(normalize!(v) - w, Inf) < eps()
```
Comment (TotalVerb) [inline test/runtests.jl]: It doesn't matter in this case, but ≈ is preferred to `@test_approx_eq` which is broken on some numeric types.

Comment (ranjanan) [inline test/runtests.jl]: @TotalVerb You're right. I guess I'll use `isapprox` instead from now on. Would you like me to change it in this PR?

Comment (KristofferC) [inline test/runtests.jl]: Better to use `@test norm(w) ≈ 1.0`

Comment (ranjanan) [inline test/runtests.jl]: @TotalVerb @KristofferC it's weird, but the test fails on Julia v0.4 using `isapprox` but passes with `@test_approx_eq`

**Comment by TotalVerb** [comment]
@ranjanan The problem you are seeing on v0.4 is not caused by this change, merely exposed by it. It is rather a leaked rounding mode introduced by a different test. [Here](https://github.com/JuliaLang/Compat.jl/pull/253/files#diff-fce720c43af3c52c862fd7451c7374b8R866), try adding the following line to reset the rounding mode, so

```julia
    setrounding(T, RoundDown)
    @test rounding(T) == RoundDown
```

becomes

```julia
    setrounding(T, RoundDown)
    @test rounding(T) == RoundDown
    setrounding(T, RoundNearest)
```

**Comment by TotalVerb** [comment]
LGTM. Thanks for the contribution!

**Pull request merged**